File: backend/app/finance.py
"""Deterministic finance math and rule-based insights.

Works against the real schema in supabase/schema.sql: transactions carry a
transaction_type ("income" | "expense") with amount always positive, and
budgets are read from the budget_progress view, which already has spent /
remaining computed server-side for the current month.

Nothing here calls the AI - it's plain arithmetic so the numbers handed to
Azure AI Foundry (and shown on the dashboard) are trustworthy.
"""

from __future__ import annotations

from collections import defaultdict


def compute_summary(transactions: list[dict]) -> dict:
    income = sum(t["amount"] for t in transactions if t["transaction_type"] == "income")
    expenses = sum(t["amount"] for t in transactions if t["transaction_type"] == "expense")
    return {
        "income": round(income, 2),
        "expenses": round(expenses, 2),
        "balance": round(income - expenses, 2),
    }


def spending_by_category(transactions: list[dict]) -> list[dict]:
    totals: dict[str, float] = defaultdict(float)
    for t in transactions:
        if t["transaction_type"] == "expense":
            totals[t.get("category", "Uncategorized")] += t["amount"]
    return [
        {"category": c, "amount": round(a, 2)}
        for c, a in sorted(totals.items(), key=lambda kv: -kv[1])
    ]


def monthly_trend(transactions: list[dict]) -> list[dict]:
    by_month: dict[str, dict] = {}
    for t in transactions:
        month = t["transaction_date"][:7]  # YYYY-MM
        entry = by_month.setdefault(month, {"month": month, "income": 0.0, "expenses": 0.0})
        if t["transaction_type"] == "income":
            entry["income"] += t["amount"]
        else:
            entry["expenses"] += t["amount"]
    return [
        {**e, "income": round(e["income"], 2), "expenses": round(e["expenses"], 2)}
        for e in sorted(by_month.values(), key=lambda e: e["month"])
    ]


def month_over_month_change(transactions: list[dict]) -> dict:
    trend = monthly_trend(transactions)
    if len(trend) < 2:
        return {"available": False}
    current, previous = trend[-1], trend[-2]
    pct_change = None
    if previous["expenses"]:
        pct_change = round(
            (current["expenses"] - previous["expenses"]) / previous["expenses"] * 100, 1
        )
    return {
        "available": True,
        "current_month": current["month"],
        "previous_month": previous["month"],
        "current_total": current["expenses"],
        "previous_total": previous["expenses"],
        "pct_change": pct_change,
    }


def generate_insights(transactions: list[dict], budgets: list[dict]) -> list[dict]:
    """budgets is expected in budget_progress shape: category, amount, spent, remaining."""
    insights: list[dict] = []

    for b in budgets:
        if b["spent"] > b["amount"]:
            insights.append(
                {
                    "type": "over_budget",
                    "severity": "warning",
                    "message": (
                        f"You're over budget in {b['category']}: "
                        f"${b['spent']:.2f} spent vs a ${b['amount']:.2f} limit."
                    ),
                }
            )
        elif b["amount"] and b["spent"] / b["amount"] > 0.9:
            insights.append(
                {
                    "type": "near_budget",
                    "severity": "info",
                    "message": (
                        f"You're close to your {b['category']} budget: "
                        f"${b['spent']:.2f} of ${b['amount']:.2f}."
                    ),
                }
            )

    momc = month_over_month_change(transactions)
    if momc["available"] and momc["pct_change"] is not None:
        if momc["pct_change"] > 15:
            insights.append(
                {
                    "type": "spending_up",
                    "severity": "warning",
                    "message": (
                        f"Spending is up {momc['pct_change']}% vs last month "
                        f"(${momc['current_total']:.2f} vs ${momc['previous_total']:.2f})."
                    ),
                }
            )
        elif momc["pct_change"] < -15:
            insights.append(
                {
                    "type": "spending_down",
                    "severity": "positive",
                    "message": f"Nice — spending is down {abs(momc['pct_change'])}% vs last month.",
                }
            )

    if not insights:
        insights.append(
            {
                "type": "all_good",
                "severity": "positive",
                "message": "No budget issues detected this month. Keep it up!",
            }
        )

    return insights

File: frontend/src/utils/finance.ts
import type { Transaction } from "../types";

// Deterministic finance math - no AI involved. Mirrors backend/app/finance.py
// so the numbers shown on screen always match what the AI is grounded on.

export interface Summary {
  income: number;
  expenses: number;
  balance: number;
}

export function computeSummary(transactions: Transaction[]): Summary {
  const income = transactions
    .filter((t) => t.transaction_type === "income")
    .reduce((s, t) => s + t.amount, 0);
  const expenses = transactions
    .filter((t) => t.transaction_type === "expense")
    .reduce((s, t) => s + t.amount, 0);
  return { income: round2(income), expenses: round2(expenses), balance: round2(income - expenses) };
}

export interface CategorySpend {
  category: string;
  amount: number;
}

export function spendingByCategory(transactions: Transaction[]): CategorySpend[] {
  const totals = new Map<string, number>();
  for (const t of transactions) {
    if (t.transaction_type === "expense") {
      totals.set(t.category, (totals.get(t.category) ?? 0) + t.amount);
    }
  }
  return [...totals.entries()]
    .map(([category, amount]) => ({ category, amount: round2(amount) }))
    .sort((a, b) => b.amount - a.amount);
}

export interface MonthTrend {
  month: string;
  income: number;
  expenses: number;
}

export function monthlyTrend(transactions: Transaction[]): MonthTrend[] {
  const byMonth = new Map<string, MonthTrend>();
  for (const t of transactions) {
    const month = t.transaction_date.slice(0, 7);
    const entry = byMonth.get(month) ?? { month, income: 0, expenses: 0 };
    if (t.transaction_type === "income") entry.income += t.amount;
    else entry.expenses += t.amount;
    byMonth.set(month, entry);
  }
  return [...byMonth.values()]
    .map((e) => ({ ...e, income: round2(e.income), expenses: round2(e.expenses) }))
    .sort((a, b) => a.month.localeCompare(b.month));
}

export interface MonthOverMonth {
  available: boolean;
  currentMonth?: string;
  previousMonth?: string;
  currentTotal?: number;
  previousTotal?: number;
  pctChange?: number | null;
}

export function monthOverMonthChange(transactions: Transaction[]): MonthOverMonth {
  const trend = monthlyTrend(transactions);
  if (trend.length < 2) return { available: false };
  const current = trend[trend.length - 1];
  const previous = trend[trend.length - 2];
  const pctChange = previous.expenses
    ? round2(((current.expenses - previous.expenses) / previous.expenses) * 100)
    : null;
  return {
    available: true,
    currentMonth: current.month,
    previousMonth: previous.month,
    currentTotal: current.expenses,
    previousTotal: previous.expenses,
    pctChange,
  };
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}
